### PR TITLE
Enhance vtk writer for cmesh and forests

### LINKED ISSUE #621: Add cmesh_vtk-writer that uses the VTK-API
Should be a similar implementation to the forest_vtk_writer. 
During the implementation try to avoid code duplication and try to reuse as much as possible from the forest_vtk_writer.

### PR DESCRIPTION
**Pull request by Davknapp**
I implemented a templated vtk writer using the vtk-API for forests and cmeshes. 
Closes #621 

many lines of codes, but a lot of them are copied -> therefore medium workload. 

**_All these boxes must be checked by the reviewers before merging the pull request:_**

As a reviewer please read through all the code lines and make sure that the code is fully understood, bug free, well-documented and well-structured.

#### General
- [ ] The reviewer executed the new code features at least once and checked the results manually

- [ ] The code follows the [t8code coding guidelines](https://github.com/holke/t8code/wiki/Coding-Guideline)
- [ ] New source/header files are properly added to the Makefiles
- [ ] The code is well documented
- [ ] All function declarations, structs/classes and their members have a proper doxygen documentation
- [ ] All new algorithms and data structures are sufficiently optimal in terms of memory and runtime (If this should be merged, but there is still potential for optimization, create a new issue)

#### Tests
- [ ] The code is covered in an existing or new test case using Google Test

#### Github action

- [ ] The code compiles without warning in debugging and release mode, with and without MPI (this should be executed automatically in a github action)
- [ ] All tests pass (in various configurations, this should be executed automatically in a github action)

  If the Pull request introduces code that is not covered by the github action (for example coupling with a new library):
  - [ ] Should this use case be added to the github action?
  - [ ] If not, does the specific use case compile and all tests pass (check manually)

#### Scripts and Wiki

- [ ] If a new directory with source-files is added, it must be covered by the `script/find_all_source_files.scp` to check the indentation of these files.
- [ ] If this PR introduces a new feature, it must be covered in an example/tutorial and a Wiki article.

#### Licence

- [ ] The author added a BSD statement to `doc/` (or already has one)

### CODE REVIEW

File: tutorials/general/t8_step1_coarsemesh.cxx [hunk 1]
```diff
@@ -76,7 +76,8 @@ t8_step1_build_tetcube_coarse_mesh (sc_MPI_Comm comm)
 static void
 t8_step1_write_cmesh_vtk (t8_cmesh_t cmesh, const char *prefix)
 {
-  t8_cmesh_vtk_write_file (cmesh, prefix);
+  //t8_cmesh_vtk_write_file (cmesh, prefix);
```
Comment (jmark) [inline tutorials/general/t8_step1_coarsemesh.cxx:79]: Why is this line commented?

**Review by jmark — CHANGES_REQUESTED** [review]
Thank you for your work!

This is a first primary review. Please address following comments before the thorough review.

- Please udpate the CMake build system accordingly.

- Is this code covered by any unit test? If not, would you at least write small test (and/or example) using this new templated style?

File: tutorials/general/t8_step2_uniform_forest.cxx [hunk 2]
```diff
@@ -100,7 +102,10 @@ t8_step2_build_uniform_forest (sc_MPI_Comm comm, t8_cmesh_t cmesh, int level)
 static void
 t8_step2_write_forest_vtk (t8_forest_t forest, const char *prefix)
 {
-  t8_forest_write_vtk (forest, prefix);
+  sc_MPI_Comm comm = t8_forest_get_mpicomm (forest);
+  //t8_forest_write_vtk (forest, prefix);
```
[the remark this answers is not in the document]
Comment (Davknapp) [reply tutorials/general/t8_step2_uniform_forest.cxx]: Undone in latest push

File: tutorials/general/t8_step2_uniform_forest.cxx [hunk 3]
```diff
@@ -44,12 +44,14 @@
...
-#include <t8.h>                                 /* General t8code header, always include this. */
-#include <t8_cmesh.h>                           /* cmesh definition and basic interface. */
-#include <t8_cmesh/t8_cmesh_examples.h>         /* A collection of exemplary cmeshes */
-#include <t8_forest/t8_forest_general.h>        /* forest definition and general interface. */
-#include <t8_forest/t8_forest_io.h>             /* forest io interface. */
+#include <t8.h>                          /* General t8code header, always include this. */
+#include <t8_cmesh.h>                    /* cmesh definition and basic interface. */
+#include <t8_cmesh/t8_cmesh_examples.h>  /* A collection of exemplary cmeshes */
+#include <t8_forest/t8_forest_general.h> /* forest definition and general interface. */
+//#include <t8_forest/t8_forest_io.h>             /* forest io interface. */
```
[the remark this answers is not in the document]
Comment (Davknapp) [reply tutorials/general/t8_step2_uniform_forest.cxx]: Undone in latest push

File: tutorials/general/t8_step1_coarsemesh.cxx (hunk 1 above)
Comment (Davknapp) [reply tutorials/general/t8_step1_coarsemesh.cxx]: Undone in latest push

**Comment by Davknapp** [comment]
> Thank you for your work!
> 
> This is a first primary review. Please address following comments before the thorough review.
> 
>     * Please udpate the CMake build system accordingly.
> 
>     * Is this code covered by any unit test? If not, would you at least write small test (and/or example) using this new templated style?

I've written a small type parametrized test for this. 
What has to be done for the CMake Buildsystem? If I understand it correctly it automatically searches for all  *.hxx files. I've only added *.hxx files so nothing should change, right?

File: src/t8_forest/t8_forest_vtk.h [hunk 4]
```diff
@@ -32,9 +32,21 @@
...
+#if T8_WITH_VTK
+#include <vtkUnstructuredGrid.h>
+#endif
+
 T8_EXTERN_C_BEGIN ();
 /* function declarations */
 
+#if T8_WITH_VTK
+void
+t8_forest_to_vtkUnstructuredGrid (t8_forest_t forest, vtkSmartPointer<vtkUnstructuredGrid> unstructuredGrid,
```
Comment (jmark) [inline src/t8_forest/t8_forest_vtk.h:44]: Is this function part of the public/internal API? Please consider to add documentation to this function declaration.

**Review by jmark — CHANGES_REQUESTED** [review]
Thank you for the nice code contribution!

As a final touch:
Would you please go through all the code comments and fix punctuation in order to fulfill our coding style guidelines?

https://github.com/DLR-AMR/t8code/wiki/Coding-Guideline#comments

**Comment by jmark** [comment]
> > Thank you for your work!
> > This is a first primary review. Please address following comments before the thorough review.
> > ```
> > * Please udpate the CMake build system accordingly.
> > 
> > * Is this code covered by any unit test? If not, would you at least write small test (and/or example) using this new templated style?
> > ```
> 
> I've written a small type parametrized test for this. What has to be done for the CMake Buildsystem? If I understand it correctly it automatically searches for all *.hxx files. I've only added *.hxx files so nothing should change, right?

Yes, indeed!

**Comment by jmark** [comment]
@Davknapp I just checked the follow-up PR https://github.com/DLR-AMR/t8code/pull/1145

Judging by the planned changes there I suggest you merge #1145 into this PR. This would save work for you and for me.

File: src/CMakeLists.txt [hunk 5]
```diff
@@ -148,6 +148,7 @@ target_sources( T8 PRIVATE
     t8_vtk/t8_vtk_unstructured.cxx 
     t8_vtk/t8_vtk_parallel.cxx 
     t8_vtk/t8_vtk_reader.cxx 
+    t8_vtk/t8_vtk_writer.cxx
```
Comment (jmark) [inline src/CMakeLists.txt:151]:
```suggestion
```

**Review by jmark — CHANGES_REQUESTED** [review]
Just minor things left. Besides that, nice work!

Would you also add the documentation to the c-interface as requested in another comment?

File: src/t8_forest/t8_forest_vtk.h (hunk 4 above)
Comment (Davknapp) [reply src/t8_forest/t8_forest_vtk.h:44]: A short sidenote: This isn't introduced by this PR, the function existed befor: Nevertheless I wrote it and didn't document it, so I will do so ;)

**Review by jmark — APPROVED** [review]

**Pull request merged**